sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -1055,43 +1059,89 @@ case class GpuCast(
 
   private def castDecimalToDecimal(input: ColumnVector,
       from: DecimalType,
-      to: DecimalType): ColumnVector = {
+      to: DecimalType,
+      ansiMode: Boolean): ColumnVector = {
 
-    // At first, we conduct overflow check onto input column.
-    // Then, we cast checked input into target decimal type.
-    val checkedInput = if (to.scale <= from.scale) {
-      // No need to promote precision unless target scale is larger than the source one,
-      // which indicates the cast is always valid when to.scale <= from.scale.
-      input.incRefCount()
-    } else {
+    val isFrom32Bit = DecimalType.is32BitDecimalType(from)
+    val isTo32Bit = DecimalType.is32BitDecimalType(to)
+    val cudfDecimal = DecimalUtil.createCudfDecimal(to.precision, to.scale)
+
+    def castCheckedDecimal(checkedInput: ColumnVector): ColumnVector = {
+      to.scale - from.scale match {
+        case 0 =>
+          if (isFrom32Bit == isTo32Bit) {
+            checkedInput.incRefCount()
+          } else {
+            // the input is already checked, just cast it
+            checkedInput.castTo(cudfDecimal)
+          }
+        case diff if diff > 0 =>
+          checkedInput.castTo(cudfDecimal)
+        case _ =>
+          withResource(checkedInput.round(to.scale, ai.rapids.cudf.RoundMode.HALF_UP)) {
+            rounded => rounded.castTo(cudfDecimal)
+          }
+      }
+    }
+
+    def checkForOverflow: ColumnVector = {
       // Check whether there exists overflow during promoting precision or not.
       // We do NOT use `Scalar.fromDecimal(-to.scale, math.pow(10, 18).toLong)` here, because
       // cuDF binaryOperation on decimal will rescale right input to fit the left one.
       // The rescaling may lead to overflow.
-      val absBound = math.pow(10, DType.DECIMAL64_MAX_PRECISION + from.scale - to.scale).toLong
-      if (ansiMode) {
+      // absBound is the maximum value that the input column can have before being casted
+      val prec = to.precision + from.scale - to.scale
+
+      // if target precision is greater than and smaller than the max/min precision that can
+      // be held in the input, go ahead with the cast without further checking
+      // When we support 128 bit Decimals we should add a check for that
+      // if (isFrom32Bit && prec > Decimal.MAX_INT_DIGITS ||
+      // !isFrom32Bit && prec > Decimal.MAX_LONG_DIGITS)
+      if (isFrom32Bit && prec > Decimal.MAX_INT_DIGITS) {
+        return input.incRefCount()
```

Comment:
I really don't like the early return here. It is inside a nested method which makes it confusing from the code what point is being returned from what method. It might be best to either refactor the code so that `checkForOverflow` became something like `checkForDecimalOverflow` and was a stand alone method, or make it so we don't return early.